pkg/webhook/health_check.go

```diff
@@ -4,30 +4,38 @@ import (
 	"bytes"
 	"crypto/tls"
 	"fmt"
+	"io"
 	"net/http"
 	"path/filepath"
 
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-// disabling gosec linting here as the http client used in this checking is intended to skip CA verification
-//
-//nolint:gosec
-var tr = &http.Transport{
-	TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
-}
-var insecureClient = &http.Client{Transport: tr}
-
 var tlsCheckerLog = logf.Log.WithName("webhook-tls-checker")
 
 func NewTLSChecker(certDir string, port int) func(*http.Request) error {
+	//nolint:forcetypeassert
+	tr := http.DefaultTransport.(*http.Transport).Clone()
+	// disabling gosec linting here as the http client used in this checking is intended to skip CA verification
+	//
+	//nolint:gosec
+	tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	// disable keep alives to ensure that http connection aren't reused, otherwise the check may
+	// fail if the cert was rotated in between
+	tr.DisableKeepAlives = true
```

Comment:
would keep alives here just cause some network flakiness? iow, if checks retry once the certs have been rotated, won't this not be a problem?

Reply:
I don't think it would be just flakiness. I didn't get the time to validate my theory, but here is the scenario that I'm thinking (with keepalive enabled):

- First health check pass, with certificate A. Because we drained the body before closing it, the connection is kept open (that's the default behaviour of the Go HTTP client, as long as the server also allow keepalive)
- Then, the certificate is renewed. Certificate A is replaced by certificate B on file, and the transport layer of the HTTP server now use certificate B for new connections.
- A new health check is started. The previous connection is open, in the pool of connections. The Go HTTP client select it and make a new request. We get a response. When we check the response, we compare what we have on disk (certificate B) with the peer cert associated to the connection (certificate A). The check fail.
- This repeat indefinitely until the connection is closed, which may happen too late, because we are always properly draining and closing the body, so the HTTP client should always keep the connection open.

Reply:
That seems likely. IIRC certs are only used for negotiating a session key, so rotating a cert wouldn't necessarily break a pre-existing connection.